src/hotspot/share/opto/vectornode.hpp

```diff
@@ -1688,15 +1688,17 @@ class SelectFromTwoVectorNode : public VectorNode {
   virtual int Opcode() const;
 };
 
-
+// The machine may not directly support the rearrange operation of an element type. In those cases,
+// we can transform the rearrange into a different element type. For example, on x86 before AVX512,
+// there is no rearrange instruction for short elements, what we will then do is to transform the
+// shuffle vector into 1 that we can do byte rearrange such that it would provide the same result.
+// This can be done in VectorRearrangeNode during code emission but we eagerly expand out this
+// because it is often the case that an index vector is reused in many rearrange operations.
```

Comment:
Thanks for this explanation!

`This can be done in VectorRearrangeNode` -> **could have** be done, because we now don't do it, right?
What do you mean by `expand out this`? Do you mean we have a separate dedicated node, so that it could possibly fold away with other nodes, such as index vector?

Reply:
Thanks for the reviews, I have added another comment sentence. Basically, by separating into a dedicated node, the preparation of the index vector can be GVN-ed across multiple rearrange operations as well as hoisted out of loops if the index vector is a loop invariant.